Expectations, part 2 (#249)
Kattis/problem-package-format: spec/2023-07-draft.md

## Example Submissions

Correct and incorrect solutions to the problem are provided in subdirectories of `submissions/`.
By default, the possible subdirectories are as in the table below, but they can
be customized, see [Default directories](#default-directories).

| Value               | Requirement                                                                           | Comment                                                   |
|---------------------|---------------------------------------------------------------------------------------|-----------------------------------------------------------|
| accepted            | Accepted as a correct solution for all test cases.                                    | At least one is required. Used to lower bound time limit. |
| rejected            | At least one case is not accepted.                                                    |                                                           |
| wrong_answer        | At least one case is wrong answer, and all cases are either wrong answer or accepted. |                                                           |

> | time_limit_exceeded | Too slow on at least one case, and all cases are either too slow or accepted.         | Used to upper bound time limit.                           |

Reviewer: I still would like to allow WA, but if nobody agrees with me, so be it.

> | accepted            | Accepted as a correct solution for all test cases.                                    | At least one is required. Used to lower bound time limit. |
> | rejected            | At least one case is not accepted.                                                    |                                                           |
> | wrong_answer        | At least one case is wrong answer, and all cases are either wrong answer or accepted. |                                                           |
> | time_limit_exceeded | Too slow on at least one case, and all cases are either too slow or accepted.         | Used to upper bound time limit.                           |

Reviewer 2: ```suggestion
| accepted            | Accepted as a correct solution for all test cases.                                    | At least one is required. Used to lower bound the time limit. |
| rejected            | At least one case is not accepted.                                                    |                                                           |
| wrong_answer        | At least one case is wrong answer, and all cases are either wrong answer or accepted. |                                                           |
| time_limit_exceeded | Too slow on at least one case, and all cases are either too slow or accepted.         | Used to upper bound the time limit.                       |
```

| required           | String or sequence of strings | `[AC, WA, TLE, RTE]`                                                   | At least one test case must have a verdict is shit subset of `AC`, `WA`, `TLE`, `RTE`. Must be a subset of `permitted`. |
| score              | Float or list of two floats   |                                                                        | The score of the submission equals the given number, or is in the given range. Only for scoring problems.               |
| message            | String                        | Empty string                                                           | This must appear as a substring in at least one `judgemessage.txt`.                                                     |
| use_for_time_limit | Bool                          | See below.                                                             | Controls whether this submission is used to determine the time limit.                                                   |

Every submission matched by the glob must:

> - only have verdicts present in `permitted`;

Reviewer: "all testcases must have verdicts from `permitted`".

> - get at least one verdict in `required`;

Reviewer: "at least some testcase must have a verdict from `required`"

- if given, get a score in the given range or equalling the given score;
- include the `message` string as a case-sensitive substring in the `judgemessage.txt` for at least one test case. 

### Groups
The `permitted`, `required`, `score`, `message`, and `use_for_time_limit`
requirements can also be given for only a subset of test cases, by adding them
under a key with the name of a test group (relative to `data/`). In this case,
the `permitted`, `required`, `message`, and `use_for_time_limit` keys only apply
to the set of test cases (recursively) in the given group.

> The `score` key puts a constraint of the aggregated score of a given test group,

Reviewer 2: ```suggestion
The `score` key puts a constraint on the aggregated score of a given test group,
```

    required: TLE
```

#### Globs

Globs can be used to apply restrictions to a subset of submissions.
It is also possible to use globs to put restrictions on a subset of test
cases, for example when test groups are not used:
```yaml
time_lmit_exceeded/solves_easy_cases.py:
  sample:
    permitted: AC
  secret/*-easy:
    permitted: AC
  secret/*-hard:
    permitted: [AC, TLE]
    required: TLE
```
This means that the submission must solve all samples and all easy cases, but
must time out on at least one of the hard cases.

Submission globs are matched against all paths to files and directories of
submissions inside and relative to the `submissions/` directory.
Test case globs are matched against all paths of test groups and test cases relative to
`submissions` and `data/` respectively (, excluding the trailing `.in`. Wildcards (`*`) only match within a file
name (e.g., do not match `/`). A test case is matched by the glob if either
itself or any of its parent test groups is matched by it, and similarly a
submission is matched if either itself or a parent directory is matched.

Using `**` to match any number of directories and `[xyz]` to match only a subset

> of characters is no supported.

Reviewer: ```suggestion
of characters is not supported.
```

> TODO: Did we allow `{a,b}.py` in the end?

Reviewer: It would be nice.

### Default directories
By default, the following requirements are defined:
```yaml
# All cases must be accepted.
accepted:
  permitted: AC
# At least one case is not accepted.
rejected:
  required: [RTE, TLE, WA]
# All cases AC or WA, at least one WA.
wrong_answer:
  permitted: [AC, WA]
  required: WA
# All cases AC or TLE, at least one TLE.
time_limit_exceeded:
  permitted: [AC, TLE]
  required: TLE
# All cases AC or RTE, at least one RTE.
run_time_error:
  permitted: [AC, RTE]
  required: RTE
# Must not WA, but fail at least once. TODO: name up for debate.
bruteforce:
  permitted: [AC, RTE, TLE]

>   required: [RTE, TLE]

Reviewer: I think this should be allowed to be `AC`.

Author: as discussed, we leave it as is. Tooling can suggest to move to `accepted/` instead.

```

> These requirements may be too strict. To e.g. allow TLE

Reviewer: ```suggestion
To e.g. allow TLE
```

Every submission matched by the glob must satisfy:
- all test cases must have only verdicts present in `permitted`;
- at least one test case must have a verdict in `required`;
- if given, the score must be in the given range or equal the given score;
- if given, the `message` string must be included as a case-sensitive substring in the `judgemessage.txt` for at least one test case. 

The tooling should check the constraints for consistency, such as that two
disjoint `permitted` sets are applied to a single `(submission, testcase)` pair.

### Groups
The `permitted`, `required`, `score`, `message`, and `use_for_time_limit`
requirements can also be given for only a subset of test cases, by adding them
under a key with the name of a test group (relative to `data/`). In this case,
the `permitted`, `required`, `message`, and `use_for_time_limit` keys only apply
to the set of test cases (recursively) in the given group.

The `score` key puts a constraint on the aggregated score of a given test group,

> _not_ on the _set_ of test cases the group contains.

Reviewer 2: This doesn't specify how a score should be aggregated on a set of test cases (which is not exactly a test group).

For example, the configuration below tests that the submission solves all cases
in `group1`, but times out on at least one case in `group2`.
```yaml
solves_group_1.py:
  sample:
    permitted: AC
  secret/group1:
    permitted: AC
  secret/group2:
    permitted: [AC, TLE]
    required: TLE
```

#### Globs

Globs can be used to apply restrictions to a subset of submissions.
It is also possible to use globs to put restrictions on a subset of test
cases, for example when test groups are not used:
```yaml
time_lmit_exceeded/solves_easy_cases.py:
  sample:
    permitted: AC
  secret/*-easy:
    permitted: AC
  secret/*-hard:
    permitted: [AC, TLE]
    required: TLE
```
This means that the submission must solve all samples and all easy cases, but
must time out on at least one of the hard cases.

Submission globs are matched against all paths to files and directories of
submissions inside and relative to the `submissions/` directory.
Test case globs are matched against all paths of test groups and test cases relative to

> `submissions` and `data/` respectively (, excluding the trailing `.in`. Wildcards (`*`) only match within a file

Reviewer 2: ```suggestion
`data/` (, excluding the trailing `.in`. Wildcards (`*`) only match within a file
```

Using `**` to match any number of directories and `[xyz]` to match only a subset
of characters is not supported. Brace expansion _is_ supported for both
submissions and test cases. Thus, one can
write `{simple,complex}.py` or `author.{py,cpp}` to match multiple files.

### Default directories
By default, the following requirements are defined:
```yaml
# All cases must be accepted.
accepted:
  permitted: AC
# At least one case is not accepted.
rejected:
  required: [RTE, TLE, WA]
# All cases AC or WA, at least one WA.
wrong_answer:
  permitted: [AC, WA]
  required: WA
# All cases AC or TLE, at least one TLE.
time_limit_exceeded:
  permitted: [AC, TLE]
  required: TLE
# All cases AC or RTE, at least one RTE.
run_time_error:
  permitted: [AC, RTE]
  required: RTE
# Must not WA, but fail at least once.
# Note that by default these are not used for determining the time limit.
bruteforce:
  permitted: [AC, RTE, TLE]
  required: [RTE, TLE]
```

The defaults can be overwritten in the `submissions.yaml` file by simply specifying the name of the directory.
Keys that are not specified are inherited from the default configuration above.

> (This is mostly supported for backwards compatibility, and not recommended for
> normal usage.)

Reviewer 2: ```suggestion
This is supported for backwards compatibility, and not recommended for normal usage.
```

```yaml
time_limit_exceeded:
  permitted: [AC, WA, TLE]
  required: TLE
```

> (Note that using `time_limit_exceeded/*` instead would impose an _additional_
> requirement, instead of _replacing_ the original requirement.)

Reviewer 2: ```suggestion
Note that the glob `time_limit_exceeded/*` would impose an _additional_
requirement, instead of _replacing_ the original requirement.
```

### Timelimit inference
Any submission that must satisfy a `required: TLE` requirement, i.e., must `TLE`
on at least one test case, is used to provide an _upper bound_ on the time limit.
Precisely, the time limit must be at most `T / time_limit_to_tle`, where `T` is
the slowest runtime over the set of test cases to which the rule applies.

> (Note that this excludes submissions that e.g. have `required: [TLE, RTE]`.)

Reviewer 2: ```suggestion
Note that this excludes submissions that e.g. have `required: [TLE, RTE]`.
```

Any submission that is not permitted to get `TLE` at all (on some subset of cases), i.e., must satisfy a
`permitted:` rule that does not contain `TLE`, is used to provide a _lower
bound_ on the time limit. Precisely, the time limit must be at least `T *
ac_to_time_limit`, where `T` is the slowest runtime over the set of test cases to
which the rule applies.

> To _opt out_ of a (set of) submission(s) from influencing the time limit, set
> `use_for_time_limit: false` alongside the `permitted:` or `required:` key that
> satisfies the constraints above.

Reviewer 2: ```suggestion
To _opt out_ of a glob for submission(s) and optional subset of testcases from influencing the time limit, set
`use_for_time_limit: false` alongside the `permitted:` and/or `required:` keys.
Note that this means that if you want to exclude a submission completely,
then you must add `use_for_time_limit: false` to every glob that matches that submission.
```

To explicitly _opt in_ a (set of) submissions(s) to be used for determining the time limit,
use `use_for_time_limit: lower` and `use_for_time_limit: upper`.
The first is equivalent to a `permitted: [AC, WA, RTE]` constraint, and the
second to a `required: [TLE]` constraint. The system may warn when this makes

> other constraints redundant.

Reviewer 2: ```suggestion
other constraints redundant and should error when it is inconsistent with other constraints.
```

be customized, see [Default directories](#default-directories).

| Value                 | Requirement                                                                           | Comment                                                       |
|-----------------------|---------------------------------------------------------------------------------------|---------------------------------------------------------------|
| `accepted`            | Accepted as a correct solution for all test cases.                                    | At least one is required. Used to lower bound the time limit. |
| `rejected`            | At least one case is not accepted.                                                    |                                                               |
| `wrong_answer`        | At least one case is wrong answer, and all cases are either wrong answer or accepted. | Used to lower bound the time limit.                           |
| `time_limit_exceeded` | Too slow on at least one case, and all cases are either too slow or accepted.         | Used to upper bound the time limit.                           |

> | `run_time_error`      | Crashes for at least one case, and all cases either crash or are accepted.            | Used to lower bound the time limit.                           |

Reviewer 3: Shouldn't `bruteforce` be added to this table? (See "Default directories")
